### PR TITLE
[MERGE][REF] sale,website_sale*: combination info logic

### PR DESCRIPTION
**Pull request by Feyensv**
Clean, factorize and optimize the code related to product variants display and price computation, especially the whole `combination_info` logic, which can now be moved to `website_sale` since the new product configurator used for backend doesn't rely on it anymore.

Enterprise PR: https://github.com/odoo/enterprise/pull/41308

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_sale/models/product_template.py [hunk 1]
```diff
@@ -164,15 +164,16 @@ def _sort_key_variant(variant):
...
+
         pricelist and pricelist.ensure_one()
         partner_sudo = self.env.user.partner_id
         pricelist = pricelist or self.env['product.pricelist']
         currency = pricelist.currency_id or self.env.company.currency_id
-
-        # Try to fetch geoip based fpos or fallback on partner one
-        fpos_id = self.env['website']._get_current_fiscal_position_id(partner_sudo)
-        fiscal_position = self.env['account.fiscal.position'].sudo().browse(fpos_id)
+        # Can we safely cache something depending on today's date ??
```
Comment (chevalierv) [inline addons/website_sale/models/product_template.py:175]: Indeed I don't think that it's a good idea

**Review by chevalierv — APPROVED** [review]
LGTM :+1: 

I really like what you've done by merging the files for website_sale_product_configurator!

File: addons/website_sale_product_configurator/controllers/main.py [hunk 2]
```diff
@@ -1,101 +1,69 @@
...
+        product = request.env['product.product'].browse(int(product_id))
 
-        return str(order.cart_quantity)
+        return request.env['ir.ui.view']._render_template(
+            'website_sale_product_configurator.optional_product_items',
+            {
+                'product': product,
+                'parent_name': product.name,
+                'parent_combination': product.product_template_attribute_value_ids,
+                'add_qty': float(add_qty) or 1.0,
```
Comment (vava-odoo) [inline addons/website_sale_product_configurator/controllers/main.py:67]: default value of kwarg
```suggestion
                'add_qty': float(add_qty),
```

File: addons/website_sale_product_configurator/controllers/website_sale.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
```
Comment (vava-odoo) [inline addons/website_sale_product_configurator/controllers/website_sale.py:1]:
```suggestion
```

File: addons/website_sale_product_configurator/controllers/website_sale.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import json
+from odoo.http import request, route
+
```
Comment (vava-odoo) [inline addons/website_sale_product_configurator/controllers/website_sale.py:6]:
```suggestion

from odoo.http import request, route
```

File: addons/website_sale_stock/models/sale_order.py [hunk 5]
```diff
@@ -106,7 +106,8 @@ def _all_product_available(self):
             product = line.product_id
             if product.type != 'product' or product.allow_out_of_stock_order:
                 continue
-            combination_info = product._get_combination_info_variant(add_qty=line.product_uom_qty)
-            if combination_info['free_qty'] == 0:
+            website = line.order_id.website_id
```
Comment (vava-odoo) [inline addons/website_sale_stock/models/sale_order.py:109]:
```suggestion
            website = self.website_id
```

File: addons/website_sale_stock/tests/test_website_sale_stock_product_warehouse.py [hunk 6]
```diff
@@ -1,10 +1,11 @@
 # -*- coding: utf-8 -*-
```
Comment (vava-odoo) [inline addons/website_sale_stock/tests/test_website_sale_stock_product_warehouse.py:1]:
```suggestion
```

**Review by vava-odoo — APPROVED** [review]
Hi @Feyensv 
Good job on this one. Looks fine to me, even if I have (very) few comments :eyes: 
Cheers!

File: addons/website_sale/models/product_template.py [hunk 7]
```diff
@@ -234,101 +234,300 @@ def _get_sales_prices(self, pricelist):
...
+
+            - has_discounted_price: True if the pricelist discount policy says
+                the price does not include the discount and there is actually a
+                discount applied (price < list_price), else False
+        """
+        self.ensure_one()
+
+        combination = combination or self.env['product.template.attribute.value']
+        parent_combination = parent_combination or self.env['product.template.attribute.value']
+        website = self.env['website'].get_current_website().with_context(self.env.context)
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/website_sale/models/product_template.py:314]: Nope, we want the returned website to have the same context as the `_get_combination_info` call, this order is on purpose :D

File: addons/website_sale/models/product_template.py [hunk 8]
```diff
@@ -234,101 +234,300 @@ def _get_sales_prices(self, pricelist):
...
-                base_unit_price=product.base_unit_count and list_price / product.base_unit_count,
-                price=price,
-                list_price=list_price,
-                price_extra=price_extra,
-                has_discounted_price=has_discounted_price,
-                prevent_zero_price_sale=prevent_zero_price_sale,
-                compare_list_price=compare_list_price,
+                    product_taxes,
+                    taxes,
+                    product_or_template,
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/website_sale/models/product_template.py:448]: too long

**Comment by Feyensv** [comment]
@robodoo r+ rebase-ff

**Comment by Feyensv** [comment]
@robodoo rebase-merge

**Comment by Feyensv** [comment]
@robodoo r+

**Pull request closed without merging**